triggers/daml/Daml/Trigger/LowLevel.daml

```diff
@@ -244,3 +259,94 @@ data Commands = Commands
   { commandId : CommandId
   , commands : [Command]
   }
+
+newtype StateT s m a = StateT { runStateT : s -> m (a, s) }
+  deriving Functor
+
+liftStateT : Functor m => m a -> StateT s m a
+liftStateT ma = StateT $ \s -> (,s) <$> ma
+
+instance Action m => Applicative (StateT s m) where
+  pure a = StateT (\s -> pure (a, s))
+  f <*> x = f >>= (<$> x)
+  
+instance Action m => Action (StateT s m) where
+  StateT x >>= f = StateT $ \s -> do
+    (x', s') <- x s
+    runStateT (f x') s'
+
+class ActionState s m | m -> s where
+  get : m s
+  put : s -> m ()
+
+modify : (Action m, ActionState s m) => (s -> s) -> m ()
+modify f = put . f =<< get
+
+instance Applicative m => ActionState s (StateT s m) where
+  get = StateT $ \s -> pure (s, s)
+  put s = StateT $ const $ pure ((), s)
+
+execStateT : Functor m => StateT s m a -> s -> m s
+execStateT (StateT fa) = fmap snd . fa
+
+zoom : Functor m => (t -> s) -> (t -> s -> t) -> StateT s m a -> StateT t m a
+zoom r w (StateT smas) = StateT $ \t ->
+  smas (r t) <&> \(a, s) -> (a, w t s)
+
+-- Must be kept in sync with Runner#handleStepFreeResult
+data TriggerF a =
+  GetTime (Time -> a)
+  | Submit (Commands, () -> a)
+  deriving Functor
+
+newtype TriggerSetup a = TriggerSetup { runTriggerSetup : Free TriggerF a }
+  deriving (Functor, Applicative, Action)
+
+newtype TriggerRule s a = TriggerRule { runTriggerRule : StateT s (Free TriggerF) a }
+  deriving (Functor, Applicative, Action)
+
+-- | Run a rule without running it.  May lose information from the rule;
```

Comment:
What exactly do you mean by “May lose information”?

Reply:
You can't take the result of `simulateRule` and get back the original rule; for example, you might have a TriggerF case that's supposed to organize a parade and pass the number of attendees to the continuation, but `simulateRule` slacks off on the logistics and just passes 42.

Reply:
Makes sense :+1: